# CNN-Modèle

Avec nos données, on peut maintenant passer à la partie modélisation

In [1]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Reshape
import pandas as pd
import numpy as np

In [2]:
X = pd.read_csv('D:/Données/ENSAE/2A/S2/Séminaire de modélisation statistique/X.csv')
Y = pd.read_csv('D:/Données/ENSAE/2A/S2/Séminaire de modélisation statistique/Y.csv')
#X_and_Y = pd.read_csv('D:/Données/ENSAE/2A/S2/Séminaire de modélisation statistique/X_and_Y.csv')

In [3]:
X = X.drop('Unnamed: 0', axis=1)
Y =Y.drop('Unnamed: 0', axis=1)
#X_and_Y = X_and_Y.drop('Unnamed: 0', axis=1)

In [4]:
X

engine_serial_number  engine_series  cycles_counter  config_A  \
0                   0.101545            0.0        0.487290  0.333333   
1                   0.101545            0.0        0.488333  0.333333   
2                   0.101545            0.0        0.489767  0.333333   
3                   0.101545            0.0        0.491983  0.333333   
4                   0.101545            0.0        0.488984  0.333333   
...                      ...            ...             ...       ...   
661595              0.000000            0.0        0.048885  0.000000   
661596              0.000000            0.0        0.135315  0.000000   
661597              0.000000            0.0        0.088646  0.000000   
661598              0.000000            0.0        0.017599  0.000000   
661599              0.000000            0.0        0.071699  0.000000   

        config_B  Interpolate_flight_leg_hours  Interpolate_SV_rank  \
0            1.0                      0.227009             0.333333   
1            1.0                      0.152526             0.333333   
2            1.0                      0.177316             0.333333   
3            1.0                      0.098576             0.333333   
4            1.0                      0.220868             0.333333   
...          ...                           ...                  ...   
661595       0.0                      0.117526             0.000000   
661596       0.0                      0.086690             0.000000   
661597       0.0                      0.086539             0.000000   
661598       0.0                      0.095543             0.000000   
661599       0.0                      0.118638             0.000000   

        Interpolate_WW_rank  Interpolate_Config_B_rank  Interpolate_var_env_1  \
0                       8.0                        0.6               0.012105   
1                       8.0                        0.6               0.001992   
2                       8.0                        0.6               0.003743   
3                       8.0                        0.6               0.000152   
4                       8.0                        0.6               0.000970   
...                     ...                        ...                    ...   
661595                  0.0                        0.0               0.003520   
661596                  0.0                        0.0               0.008363   
661597                  0.0                        0.0               0.019375   
661598                  0.0                        0.0               0.000963   
661599                  0.0                        0.0               0.001441   

        Interpolate_var_env_2  Interpolate_var_env_3  Interpolate_var_env_4  \
0                    0.008404               0.435294                    0.0   
1                    0.008404               0.588235                    0.0   
2                    0.001931               0.670588                    0.0   
3                    0.008404               0.529412                    0.0   
4                    0.002376               0.611765                    0.0   
...                       ...                    ...                    ...   
661595               0.019520               0.568177                    0.0   
661596               0.000672               0.388235                    0.0   
661597               0.026428               0.494118                    0.0   
661598               0.028144               0.717647                    0.0   
661599               0.026428               0.635294                    0.0   

        Interpolate_var_env_5  id_int  
0                    0.198413  1823.0  
1                    0.085317  1823.0  
2                    0.107143  1823.0  
3                    0.107143  1823.0  
4                    0.111111  1823.0  
...                       ...     ...  
661595               0.150087  8438.0  
661596               0.111111  8438.

In [5]:
X.dtypes

engine_serial_number            float64
engine_series                   float64
cycles_counter                  float64
config_A                        float64
config_B                        float64
Interpolate_flight_leg_hours    float64
Interpolate_SV_rank             float64
Interpolate_WW_rank             float64
Interpolate_Config_B_rank       float64
Interpolate_var_env_1           float64
Interpolate_var_env_2           float64
Interpolate_var_env_3           float64
Interpolate_var_env_4           float64
Interpolate_var_env_5           float64
id_int                          float64
dtype: object

In [6]:
Y

Interpolate_egt_slope  id_int
0                   -0.027761    1823
1                   -0.027761    1823
2                   -0.027761    1823
3                   -0.027761    1823
4                   -0.027761    1823
...                       ...     ...
661595              -0.029193    8438
661596              -0.029193    8438
661597              -0.029193    8438
661598              -0.029193    8438
661599              -0.029193    8438

[661600 rows x 2 columns]

In [7]:
Y.dtypes

Interpolate_egt_slope    float64
id_int                     int64
dtype: object

### Multiple Input Series

In [8]:
# multivariate cnn example
from numpy import array
from numpy import hstack
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Flatten
from keras.layers.convolutional import Conv1D
from keras.layers.convolutional import MaxPooling1D

In [9]:
X.columns

Index(['engine_serial_number', 'engine_series', 'cycles_counter', 'config_A',
       'config_B', 'Interpolate_flight_leg_hours', 'Interpolate_SV_rank',
       'Interpolate_WW_rank', 'Interpolate_Config_B_rank',
       'Interpolate_var_env_1', 'Interpolate_var_env_2',
       'Interpolate_var_env_3', 'Interpolate_var_env_4',
       'Interpolate_var_env_5', 'id_int'],
      dtype='object')

Les deux cellules ci-dessous, je vais le refaire au propre après

## Découpage par intervalles de 100

In [10]:
print("on a",Y.shape[0]/100, "intervalles.")

on a 6616.0 intervalles.


In [11]:
print(round(0.75*Y.shape[0]/100)) #On va prendre 5018 intervalles pour le train et 
print(round(0.25*Y.shape[0]/100)) # 1673 pour le test
print(round(0.75*Y.shape[0]/100)+ round(0.25*Y.shape[0]/100)) #On vérifie qu'on a bien la totalité des intervalles

4962
1654
6616


In [13]:
X

engine_serial_number  engine_series  cycles_counter  config_A  \
0                   0.101545            0.0        0.487290  0.333333   
1                   0.101545            0.0        0.488333  0.333333   
2                   0.101545            0.0        0.489767  0.333333   
3                   0.101545            0.0        0.491983  0.333333   
4                   0.101545            0.0        0.488984  0.333333   
...                      ...            ...             ...       ...   
661595              0.000000            0.0        0.048885  0.000000   
661596              0.000000            0.0        0.135315  0.000000   
661597              0.000000            0.0        0.088646  0.000000   
661598              0.000000            0.0        0.017599  0.000000   
661599              0.000000            0.0        0.071699  0.000000   

        config_B  Interpolate_flight_leg_hours  Interpolate_SV_rank  \
0            1.0                      0.227009             0.333333   
1            1.0                      0.152526             0.333333   
2            1.0                      0.177316             0.333333   
3            1.0                      0.098576             0.333333   
4            1.0                      0.220868             0.333333   
...          ...                           ...                  ...   
661595       0.0                      0.117526             0.000000   
661596       0.0                      0.086690             0.000000   
661597       0.0                      0.086539             0.000000   
661598       0.0                      0.095543             0.000000   
661599       0.0                      0.118638             0.000000   

        Interpolate_WW_rank  Interpolate_Config_B_rank  Interpolate_var_env_1  \
0                       8.0                        0.6               0.012105   
1                       8.0                        0.6               0.001992   
2                       8.0                        0.6               0.003743   
3                       8.0                        0.6               0.000152   
4                       8.0                        0.6               0.000970   
...                     ...                        ...                    ...   
661595                  0.0                        0.0               0.003520   
661596                  0.0                        0.0               0.008363   
661597                  0.0                        0.0               0.019375   
661598                  0.0                        0.0               0.000963   
661599                  0.0                        0.0               0.001441   

        Interpolate_var_env_2  Interpolate_var_env_3  Interpolate_var_env_4  \
0                    0.008404               0.435294                    0.0   
1                    0.008404               0.588235                    0.0   
2                    0.001931               0.670588                    0.0   
3                    0.008404               0.529412                    0.0   
4                    0.002376               0.611765                    0.0   
...                       ...                    ...                    ...   
661595               0.019520               0.568177                    0.0   
661596               0.000672               0.388235                    0.0   
661597               0.026428               0.494118                    0.0   
661598               0.028144               0.717647                    0.0   
661599               0.026428               0.635294                    0.0   

        Interpolate_var_env_5  id_int  
0                    0.198413  1823.0  
1                    0.085317  1823.0  
2                    0.107143  1823.0  
3                    0.107143  1823.0  
4                    0.111111  1823.0  
...                       ...     ...  
661595               0.150087  8438.0  
661596               0.111111  8438.

In [16]:
XY_sample = X.groupby(["id_int"]).count()
XY_sample = XY_sample.sample(5018)
XY_sample = XY_sample.reset_index()
XY_sample

id_int  engine_serial_number  engine_series  cycles_counter  config_A  \
0     6875.0                   100            100             100       100   
1     6263.0                   100            100             100       100   
2     7065.0                   100            100             100       100   
3     3420.0                   100            100             100       100   
4     8026.0                   100            100             100       100   
...      ...                   ...            ...             ...       ...   
5013  3366.0                   100            100             100       100   
5014  2769.0                   100            100             100       100   
5015  4656.0                   100            100             100       100   
5016  6476.0                   100            100             100       100   
5017  3481.0                   100            100             100       100   

      config_B  Interpolate_flight_leg_hours  Interpolate_SV_rank  \
0          100                           100                  100   
1          100                           100                  100   
2          100                           100                  100   
3          100                           100                  100   
4          100                           100                  100   
...        ...                           ...                  ...   
5013       100                           100                  100   
5014       100                           100                  100   
5015       100                           100                  100   
5016       100                           100                  100   
5017       100                           100                  100   

      Interpolate_WW_rank  Interpolate_Config_B_rank  Interpolate_var_env_1  \
0                     100                        100                    100   
1                     100                        100                    100   
2                     100                        100                    100   
3                     100                        100                    100   
4                     100                        100                    100   
...                   ...                        ...                    ...   
5013                  100                        100                    100   
5014                  100                        100                    100   
5015                  100                        100                    100   
5016                  100                        100                    100   
5017                  100                        100                    100   

      Interpolate_var_env_2  Interpolate_var_env_3  Interpolate_var_env_4  \
0                       100                    100                    100   
1                       100                    100                    100   
2                       100                    100                    100   
3                       100                    100                    100   
4                       100                    100                    100   
...                     ...                    ...                    ...   
5013                    100                    100                    100   
5014                    100                    100                    100   
5015                    100                    100                    100   
5016                    100                    100                    100   
5017                    100                    100                    100   

      Interpolate_var_env_5  
0                       100  
1                       100  
2                       100  
3                       100  
4                       100  
...                     ...  
5013                    100  
5014                    100  
5015                    100  
5016                    100  
5017                    

In [18]:
#ATTENTION : cette cellule prend 10 minutes à s'éxécuter
#Cette méthode prend plus de temps que la technique habituelle split train test,
# mais nous permet de garder chaque groupe d'intervalle ensemble

for i in XY_sample["id_int"]:
    X_train = X[X["id_int"].isin(XY_sample["id_int"])]
    Y_train = Y[Y["id_int"].isin(XY_sample["id_int"])]
    
    X_test =  X[~X["id_int"].isin(XY_sample["id_int"])]
    Y_test =  Y[~Y["id_int"].isin(XY_sample["id_int"])]
    

In [19]:
Y_test

Interpolate_egt_slope  id_int
100                 -0.025889    1824
101                 -0.025889    1824
102                 -0.025889    1824
103                 -0.025889    1824
104                 -0.025889    1824
...                       ...     ...
660795              -0.003273    8430
660796              -0.003273    8430
660797              -0.003273    8430
660798              -0.003273    8430
660799              -0.003273    8430

[159800 rows x 2 columns]

In [21]:
X_train

engine_serial_number  engine_series  cycles_counter  config_A  \
0                   0.101545            0.0        0.487290  0.333333   
1                   0.101545            0.0        0.488333  0.333333   
2                   0.101545            0.0        0.489767  0.333333   
3                   0.101545            0.0        0.491983  0.333333   
4                   0.101545            0.0        0.488984  0.333333   
...                      ...            ...             ...       ...   
661595              0.000000            0.0        0.048885  0.000000   
661596              0.000000            0.0        0.135315  0.000000   
661597              0.000000            0.0        0.088646  0.000000   
661598              0.000000            0.0        0.017599  0.000000   
661599              0.000000            0.0        0.071699  0.000000   

        config_B  Interpolate_flight_leg_hours  Interpolate_SV_rank  \
0            1.0                      0.227009             0.333333   
1            1.0                      0.152526             0.333333   
2            1.0                      0.177316             0.333333   
3            1.0                      0.098576             0.333333   
4            1.0                      0.220868             0.333333   
...          ...                           ...                  ...   
661595       0.0                      0.117526             0.000000   
661596       0.0                      0.086690             0.000000   
661597       0.0                      0.086539             0.000000   
661598       0.0                      0.095543             0.000000   
661599       0.0                      0.118638             0.000000   

        Interpolate_WW_rank  Interpolate_Config_B_rank  Interpolate_var_env_1  \
0                       8.0                        0.6               0.012105   
1                       8.0                        0.6               0.001992   
2                       8.0                        0.6               0.003743   
3                       8.0                        0.6               0.000152   
4                       8.0                        0.6               0.000970   
...                     ...                        ...                    ...   
661595                  0.0                        0.0               0.003520   
661596                  0.0                        0.0               0.008363   
661597                  0.0                        0.0               0.019375   
661598                  0.0                        0.0               0.000963   
661599                  0.0                        0.0               0.001441   

        Interpolate_var_env_2  Interpolate_var_env_3  Interpolate_var_env_4  \
0                    0.008404               0.435294                    0.0   
1                    0.008404               0.588235                    0.0   
2                    0.001931               0.670588                    0.0   
3                    0.008404               0.529412                    0.0   
4                    0.002376               0.611765                    0.0   
...                       ...                    ...                    ...   
661595               0.019520               0.568177                    0.0   
661596               0.000672               0.388235                    0.0   
661597               0.026428               0.494118                    0.0   
661598               0.028144               0.717647                    0.0   
661599               0.026428               0.635294                    0.0   

        Interpolate_var_env_5  id_int  
0                    0.198413  1823.0  
1                    0.085317  1823.0  
2                    0.107143  1823.0  
3                    0.107143  1823.0  
4                    0.111111  1823.0  
...                       ...     ...  
661595               0.150087  8438.0  
661596               0.111111  8438.

## Test en divisant en 6616 dataframe différent, avec pour chacun une prédiction

In [23]:
listOfDfs = [X_train.loc[idx] for idx in np.split(X_train.index,100)]
listOfDfs

[      engine_serial_number  engine_series  cycles_counter  config_A  config_B  \
 0                 0.101545            0.0        0.487290  0.333333       1.0   
 1                 0.101545            0.0        0.488333  0.333333       1.0   
 2                 0.101545            0.0        0.489767  0.333333       1.0   
 3                 0.101545            0.0        0.491983  0.333333       1.0   
 4                 0.101545            0.0        0.488984  0.333333       1.0   
 ...                    ...            ...             ...       ...       ...   
 6713              0.023547            0.0        0.135315  0.000000       0.0   
 6714              0.023547            0.0        0.129840  0.000000       0.0   
 6715              0.023547            0.0        0.124886  0.000000       0.0   
 6716              0.023547            0.0        0.138704  0.000000       0.0   
 6717              0.023547            0.0        0.129579  0.000000       0.0   
 
       Interpo

In [20]:
X_train.dtypes

engine_serial_number            float64
engine_series                   float64
cycles_counter                  float64
config_A                        float64
config_B                        float64
Interpolate_flight_leg_hours    float64
Interpolate_SV_rank             float64
Interpolate_WW_rank             float64
Interpolate_Config_B_rank       float64
Interpolate_var_env_1           float64
Interpolate_var_env_2           float64
Interpolate_var_env_3           float64
Interpolate_var_env_4           float64
Interpolate_var_env_5           float64
id_int                          float64
dtype: object

In [125]:
print(X_train.shape, X_test.shape, Y_train.shape, Y_test.shape)

(501800, 15) (167300, 15) (501800, 2) (167300, 2)


## Le modèle

In [129]:
# the dataset knows the number of features, e.g. 2
n_features = X_train.shape[1] # On a 15 colonnes d'input
n_features

15

In [20]:
# define model
model = Sequential()
model.add(Conv1D(filters=64, kernel_size=2, activation='relu', input_shape=(n_steps, n_features)))
model.add(MaxPooling1D(pool_size=2))
model.add(Flatten())
model.add(Dense(50, activation='relu'))
model.add(Dense(1))
model.compile(optimizer='adam', loss='mse')

In [22]:
# On fit le modèle, et ça prend  ENVIRON 1H
model.fit(X_train, y_train, epochs=10)

In [23]:
#on prédit des valeurs pour X_test
preds = model.predict(X_test)
print(preds)

[[-0.03264383]
 [-0.03264383]
 [-0.03264383]
 ...
 [-0.03264383]
 [-0.03264383]
 [-0.03264383]]


In [30]:
preds

array([[-0.03264383],
       [-0.03264383],
       [-0.03264383],
       ...,
       [-0.03264383],
       [-0.03264383],
       [-0.03264383]], dtype=float32)

In [46]:
data_preds = pd.DataFrame(preds)
data_preds.sample(40)
#On a un problème : notre modèle ne nous prédit qu'une seule valeur pour egt_slope

0
95735  -0.032644
73559  -0.032644
3398   -0.032644
115643 -0.032644
51374  -0.032644
125444 -0.032644
42081  -0.032644
14565  -0.032644
60036  -0.032644
98612  -0.032644
117664 -0.032644
115813 -0.032644
14293  -0.032644
42819  -0.032644
86394  -0.032644
134215 -0.032644
52318  -0.032644
94883  -0.032644
110262 -0.032644
55847  -0.032644
138159 -0.032644
53583  -0.032644
49805  -0.032644
15622  -0.032644
95310  -0.032644
118032 -0.032644
117881 -0.032644
36869  -0.032644
79369  -0.032644
119134 -0.032644
2336   -0.032644
117054 -0.032644
38052  -0.032644
31603  -0.032644
113274 -0.032644
126079 -0.032644
46571  -0.032644
89866  -0.032644
98996  -0.032644
84715  -0.032644

In [51]:
M=np.matrix(preds)
print(M.shape) # On a perdu 100 données

(149776, 1)


In [52]:
Y_test_2 = Y_test.head(149776)
Y_test_2 = Y_test_2[['Interpolate_egt_slope']]

In [53]:
Y_test_2

Interpolate_egt_slope
220202               0.011432
318518              -0.015750
147384              -0.025875
271519              -0.007881
38264               -0.030609
...                       ...
317566              -0.025633
365341              -0.028490
161308              -0.023947
327022               0.003361
63121               -0.066691

[149776 rows x 1 columns]

In [54]:
from sklearn.metrics import mean_squared_error
mean_squared_error(Y_test_2,preds)

0.043235160544800394

In [ ]:
Ytest=np.asanyarray(Y_test)  
Ytest=Ytest.reshape(-1,1) 
Ytest = scaler.inverse_transform(Ytest)



Maintenant, il faudrait compter le nombre de bonnes prédictions qu'on a eu, pour "noter notre modèle"